packages/functional-tests/src/app.ts

```diff
@@ -53,37 +55,94 @@ export default class App {
     private userJoined = async (user: MRESDK.User) => {
         console.log(`user-joined: ${user.name}, ${user.id}`);
 
-        this.latestUser = user;
 
         let testName: string;
         if (Array.isArray(this.params.test) && this.params.test.length > 0) {
             testName = this.params.test[0];
         } else {
             testName = this.params.test as string;
         }
-
-        await this.startTest(testName);
+        if (testName) {
+            await this.startTest(testName, user);
+            this.rpc.send('functional-test:close-connection');
+        } else {
+            if (!this.firstUser) {
+                this.runFunctionalTestMenu();
+            }
+            this.firstUser = user;
+        }
     }
-
     private userLeft = (user: MRESDK.User) => {

```

Comment:
If the last user leaves then another one joins, I think it will be broken. This means the app won’t be able to run in a long-lived way. You could handle this by clearing `firstUser` in `userLeft` if the user.id matches.

Reply:
Oh wait, I see you're overwriting firstUser. nevermind!